preload

```diff
@@ -136,56 +134,58 @@ fi
 if [ $reloadlist == 1 ];then
   echo "Generating list for '$prefix'..." 1>&2
   
-  #for app_categories() and app_status() functions
-  if ! command -v app_categories >/dev/null ;then
+  #for app_prefix_categories() and app_status() functions
+  if ! command -v app_prefix_category >/dev/null ;then
     source "${DIRECTORY}/api"
   fi
   
   if [ ! -z "$prefix" ];then
     echo "Showing apps within $prefix/" 1>&2
-    vfiles="$(app_categories | grep . | sort -f | uniq | grep "^$prefix/" | sed "s+$prefix/++g")" #generate a virtual file system with apps in folders represented as subdirectories
+    vfiles="$(app_prefix_category "$prefix" | grep . | sort -f | uniq | sed "s+$prefix/++g")" #generate a virtual file system with apps in folders represented as subdirectories
   else
-    vfiles="$(app_categories | grep . | sort -f | uniq)"
+    vfiles="$(app_prefix_category "" | grep . | sort -f | uniq | grep -v '^hidden/')"
   fi
   
   #remove apps within categories - show this layer of stuff only.
-  vfiles="$(sed 's+/.*+/+g' <<<"$vfiles" | sort -f | uniq)"
+  vfiles="$(sed 's+/.*+/+g' <<<"$vfiles" | uniq)"
   
   #get list of apps - excluding folders and apps that are incompatible with CPU architecture
   APPS="$(grep -v '/' <<<"$vfiles" | list_intersect "$(list_apps cpu_installable)")"
   
-  #get list of folders - excluding apps - and hide the hidden folder.
-  DIRS="$(grep '/' <<<"$vfiles" | tr -d '/' | grep -vFx "hidden")"
+  #get list of folders - excluding apps.
+  DIRS="$(grep '/' <<<"$vfiles" | tr -d '/')"
   
   #shuffle the list if enabled
-  if [ "$(cat "${DIRECTORY}/data/settings/Shuffle App list")" == 'Yes' ];then
+  if [ "$(while read line; do echo $line; done 2>/dev/null < "${DIRECTORY}/data/settings/Shuffle App list")" == 'Yes' ];then
```

Comment:
What is the purpose of this change? This seems like an over-complicated way to read one line from a file.

Reply:
this loop is the theoretically fastest way to read a file from bash, since while, read, and echo are builtin functions while cat is an external program which has to be loaded into memory

basically its just reducing syscalls in what I found was a "hot" codepath.

Reply:
> this loop is the theoretically fastest way to read a file from bash, since while, read, and echo are builtin functions while cat is an external program which has to be loaded into memory
> 
> basically its just reducing syscalls in what I found was a "hot" codepath.

I see. Well if that's what you're trying to do, then I'd like to introduce you to your new favorite method!
```
variable="$(< /home/pi/pi-apps/etc/categories)"
echo "$variable"
```
I figured this out by accident. 😂 I'm not sure if this usage is documented, but it appears to work.

Reply:
actually I already have another method I employed already for getting the first line of a file (so no while loop)
```
read -r 2>/dev/null line <"${DIRECTORY}/apps/${app}/description"
```

pretty similar, but no variable setting

Reply:
I've gone ahead and removed your while loops and read commands from preload. This probably reduces overhead a little, and it definitely improves readibility.

Reply:
your method is slower.... by a lot
in fact, mine is infinitly fast

```
garrett@garrett-usb:~$ time status="$(< "${DIRECTORY}/data/status/${app}")"; echo $status

real	0m0.002s
user	0m0.000s
sys	0m0.002s
installed
garrett@garrett-usb:~$ unset status
garrett@garrett-usb:~$ time read -r 2>/dev/null status <"${DIRECTORY}/data/status/${app}"; echo $status

real	0m0.000s
user	0m0.000s
sys	0m0.000s
installed
```

Reply:
(also no this is running on sd card, not a usb drive... I am too lazy to change the system name 😆 )